### DISCUSSION TITLE
Why python@3.9 is needed to install php@7.2

### DISCUSSION THREAD

**Discussion by hide-in-code** [Homebrew/discussions#2514] [Casks] [answered]
I don't think this is a bug. when I use brew to install php@7.2, brew will prompt and install python@3.9, As far as I know, php itself does not depend on anything in python, and I have installed python3.7.3 on my MAC, this Installation caused my original environment not work properly
![image](https://user-images.githubusercontent.com/19517597/142999549-fb9f2e2c-6ee9-42de-a72c-9f99771258e9.png)

--------------
I do not  really care about whether php really depends on python,  but that  php Installation of brew  will change my original python3.7 to python3.9, which will makes me can't work properly

**Comment by carlocab** [comment]
`php@7.2` depends on `glib` which depends on `python@3.9`. See `brew deps --tree php@7.2` or `brew deps --graph php@7.2`.

**Reply by hide-in-code** [reply]
Thank you for your attention.   it is found that php does depend on python by using  `brew deps`,  but I think this dependency  only occur when using brew installation, I use docker to perform a php compiling installation,  there is no such file like `*glib*`, here my records:

```
docker pull alphine
docker run -itd --name "php_test"  alpine
docker exec -it php_test /bin/sh
mkdir -p /data/ll && cd /data/ll
wget https://www.php.net/distributions/php-7.4.26.tar.gz
tar -xvf  php-7.4.26.tar.gz

apk --update --no-cache add gcc
apk --update --no-cache add make
apk --update --no-cache add libxml2
apk --update --no-cache add libxml2-dev
apk --update --no-cache add openssl
apk --update --no-cache add openssl-dev
apk --update --no-cache add curl
apk --update --no-cache add curl-dev
apk --update --no-cache add libjpeg
apk --update --no-cache add libjpeg-dev
apk --update --no-cache add libpng
apk --update --no-cache add libpng-dev
apk --update --no-cache add freetype
apk --update --no-cache add freetype-dev
apk --update --no-cache add pcre
apk --update --no-cache add pcre-dev
apk --update --no-cache add libxslt
apk --update --no-cache add libxslt-dev
apk --update --no-cache add bzip2
apk --update --no-cache add bzip2-dev
apk --update --no-cache add krb5-dev
apk --update --no-cache add krb5
apk --update --no-cache add sqlite
apk --update --no-cache add sqlite-dev
apk --update --no-cache add gettext
apk --update --no-cache add oniguruma
apk --update --no-cache add oniguruma-dev

cd  php-7.4.26
make /data/php
./configure --prefix=/data/php --with-curl --with-freetype-dir --with-gd --with-gettext --with-iconv-dir --with-kerberos --with-libdir=lib64 --with-libxml-dir --with-mysqli --with-openssl --with-pcre-regex --with-pdo-mysql --with-pdo-sqlite --with-pear --with-png-dir --with-jpeg-dir --with-xmlrpc --with-xsl --with-zlib --with-bz2 --with-mhash --enable-fpm --enable-bcmath --enable-libxml --enable-inline-optimization --enable-gd-native-ttf --enable-mbregex --enable-mbstring --enable-opcache --enable-pcntl --enable-shmop --enable-soap --enable-sockets --enable-sysvsem --enable-sysvshm --enable-xml --enable-zip
make && make install
```

![image](https://user-images.githubusercontent.com/19517597/143158506-ddad95f9-2ecf-4c2b-9938-f5610c1e01b1.png)

I am confused about this

**Reply by hide-in-code** [reply]
Actually, my  concern is not whether php really depends on python, but that python3.7 has been installed on my computer, and the installation of brew will change my original python3.7 to python3.9, which will make me can't work properly

**Reply by carlocab** [reply]
What does `brew linkage php@7.2` say?

You can do still do
```
brew unlink python@3.9
brew link --force python@3.7
```
to use Python 3.7.

**Reply by hide-in-code** [reply]
following are `brew linkage php@7.2` outputs:

    System libraries:
    /usr/lib/libSystem.B.dylib
    /usr/lib/libbz2.1.0.dylib
    /usr/lib/libc++.1.dylib
    /usr/lib/libedit.3.dylib
    /usr/lib/libexpat.1.dylib
    /usr/lib/libexslt.0.dylib
    /usr/lib/libiconv.2.dylib
    /usr/lib/libicucore.A.dylib
    /usr/lib/libncurses.5.4.dylib
    /usr/lib/libresolv.9.dylib
    /usr/lib/libxml2.2.dylib
    /usr/lib/libxslt.1.dylib
    /usr/lib/libz.1.dylib
    /usr/local/opt/tidy-html5/lib/libtidy.5.dylib
    Homebrew libraries:
    /usr/local/opt/apr/libexec/lib/libapr-1.0.dylib (apr)
    /usr/local/opt/apr-util/libexec/lib/libaprutil-1.0.dylib (apr-util)
    /usr/local/opt/argon2/lib/libargon2.1.dylib (argon2)
    /usr/local/opt/aspell/lib/libaspell.15.dylib (aspell)
    /usr/local/opt/aspell/lib/libpspell.15.dylib (aspell)
    /usr/local/opt/curl/lib/libcurl.4.dylib (curl)
    /usr/local/opt/freetds/lib/libsybdb.5.dylib (freetds)
    /usr/local/opt/freetype/lib/libfreetype.6.dylib (freetype)
    /usr/local/opt/gettext/lib/libintl.8.dylib (gettext)
    /usr/local/opt/gmp/lib/libgmp.10.dylib (gmp)
    /usr/local/opt/icu4c/lib/libicudata.69.dylib (icu4c)
    /usr/local/opt/icu4c/lib/libicui18n.69.dylib (icu4c)
    /usr/local/opt/icu4c/lib/libicuio.69.dylib (icu4c)
    /usr/local/opt/icu4c/lib/libicuuc.69.dylib (icu4c)
    /usr/local/opt/jpeg/lib/libjpeg.9.dylib (jpeg)
    /usr/local/opt/krb5/lib/libcom_err.3.0.dylib (krb5)
    /usr/local/opt/krb5/lib/libgssapi_krb5.2.2.dylib (krb5)
    /usr/local/opt/krb5/lib/libk5crypto.3.1.dylib (krb5)
    /usr/local/opt/krb5/lib/libkrb5.3.3.dylib (krb5)
    /usr/local/opt/libpng/lib/libpng16.16.dylib (libpng)
    /usr/local/opt/libpq/lib/libpq.5.dylib (libpq)
    /usr/local/opt/libsodium/lib/libsodium.23.dylib (libsodium)
    /usr/local/opt/libzip/lib/libzip.5.dylib (libzip)
    /usr/local/opt/openldap/lib/liblber-2.5.0.dylib (openldap)
    /usr/local/opt/openldap/lib/libldap-2.5.0.dylib (openldap)
    /usr/local/opt/openssl@1.1/lib/libcrypto.1.1.dylib (openssl@1.1)
    /usr/local/opt/openssl@1.1/lib/libssl.1.1.dylib (openssl@1.1)
    /usr/local/opt/sqlite/lib/libsqlite3.0.dylib (sqlite)
    /usr/local/opt/unixodbc/lib/libodbc.2.dylib (unixodbc)
    /usr/local/opt/webp/lib/libwebp.7.dylib (webp)
    Indirect dependencies with linkage:
    krb5

my solution is：

    brew uninstall --ignore-dependencies python@3.9
    brew install python@3.7

I don't know what will happen after I reinstall python

**Reply by carlocab** [reply]
If you don't need to link `python@3.7` then you don't need to uninstall `python@3.9`. Doing `brew unlink python@3.9` will be more reliable, since uninstalling can only cause you weird issues down the road.

**Reply by hide-in-code** [reply]
thank you for your reply，If possible, I hope that when brew installs or updates a new software, brew could keep the existing software environment on the target computer, so that people who use it will not feel weird

**Reply by SMillerDev** [reply]
Please also note that [Alpine Linux](https://en.wikipedia.org/wiki/Alpine_Linux) is a very different OS from most since it uses musl instead of glibc so dependencies are not always comparable.

**Reply by hide-in-code** [reply]
can I use ubuntu to test it? both compiling installation and Linuxbrew. what I want to prove is that python should not be used to install php,  it is not necessary,  otherwise people who use brew to install php will have to install another software on their computer.(for me, I need both of them, so I just need to solve multiple python versions)

**Reply by SMillerDev** [reply]
Yes, Ubuntu is a better comparison.

Regarding the multiple python versions, you might be interested to read:
https://justinmayer.com/posts/homebrew-python-is-not-for-you/

**Reply by hide-in-code** [reply]
that's it !!  I think this is the best answer for my confusion, I should not use homebrew python, I didn’t change anything … but homebrew did
![image](https://user-images.githubusercontent.com/19517597/143201939-a7ac44ba-24bf-4f95-88ea-6e8faf602dab.png)
we have no control over this
thinks a lot

**Comment by iMichka** [comment]
> Actually, my concern is not whether php really depends on python, but that python3.7 has been installed on my computer, and the installation of brew will change my original python3.7 to python3.9, which will make me can't work properly

@hide-in-code that's why you should really use `virtualenvs` with Python, to be able to have multiple versions of Python installed on your systems and be able to have multiple work environments.

**Reply by hide-in-code** [reply]
sounds like a effective way to skip this problem,  after this incident, I will consider using anaconda or pyenv to manage my python environment. (It would be great If brew can solve this problem. T^T)

**Reply by iMichka** [reply]
`pyenv` does solve it, so we do not need to solve this topic twice.
We just provide a list of Python's you can install, that's all. After that you can use the specialized tools of the language, they do it better than us.

**Reply by hide-in-code** [reply]
ok, I understand, thanks for your reply

**Comment by hide-in-code** [comment, marked as answer]
suggest answer: you shouldn’t use the system-bundled Python, don’t use Homebrew Python. It’s not meant for you.
https://github.com/Homebrew/discussions/discussions/2514#discussioncomment-1692308
https://justinmayer.com/posts/homebrew-python-is-not-for-you/

**Comment by carlocab** [comment]
I am removing the `glib` dependency from `php@7.2` in Homebrew/homebrew-core#89908.

**Reply by hide-in-code** [reply]
wow ! thanks a million ，thanks to the project team
